modules/git/repo_index.go

```diff
@@ -119,11 +119,36 @@ func (repo *Repository) RemoveFilesFromIndex(filenames ...string) error {
 	})
 }
 
+type IndexObjectInfo struct {
+	Mode     string
+	Object   ObjectID
+	Filename string
+}
+
+// AddObjectsToIndex adds the provided object hashes to the index at the provided filenames
+func (repo *Repository) AddObjectsToIndex(objects ...IndexObjectInfo) error {
+	cmd := NewCommand(repo.Ctx, "update-index", "--add", "--replace", "-z", "--index-info")
+	stdout := new(bytes.Buffer)
+	stderr := new(bytes.Buffer)
+	buffer := new(bytes.Buffer)
+	for _, object := range objects {
+		buffer.WriteString(object.Mode + " ")
+		buffer.WriteString(object.Object.String())
+		buffer.WriteByte('\t')
+		buffer.WriteString(object.Filename)
+		buffer.WriteByte('\000')
```

Comment:
Would a helper function be warrenet that accepts `buffer` and `objects` and mutates the `buffer`?

Reply:
My intuition would be to say no, since the syntax is coupled to the command line arguments. We don't need to remove from the index first because `--replace` is specified, and we use `\000` because of `-z`.

But I can make a helper function if that is preferred.

Reply:
It just seems like a bit of a odd incantation. Is there any reference on the format?

Reply:
There's this:
https://git-scm.com/docs/git-update-index#_using_index_info

But I read the git source code to check that it does the same as `--cacheinfo`. Since it wasn't obvious from that description that `--add` and `--replace` would have an effect and that the stage would be the same.

Reply:
Maybe also just write it on one line:

````go
buffer.WriteString(object.Mode + " " + object.Object.String() + '\t' + object.Filename + '\000')

Reply:
I was following what the existing remove file code was doing. I've updated both now to use the newer format and put it on a single line.

`\0` is not valid in Go the way it is in C, it has to be `\000` or `\x00`.